rules/opentitan/cc.bzl

```diff
@@ -315,11 +315,20 @@ opentitan_binary = rv_rule(
 def _opentitan_test(ctx):
     exec_env = ctx.attr.exec_env[ExecEnvInfo]
 
-    # If the test is supplied exactly one file and no deps _and_ that file
-    # is a provider for the current exec_env, then we assume that it's a
-    # pre-built binary.
-    if len(ctx.attr.srcs) == 1 and len(ctx.attr.deps) == 0 and exec_env.provider in ctx.attr.srcs[0]:
-        p = ctx.attr.srcs[0][exec_env.provider]
+    # If the test is supplied a binary and no sources, then use that binary
+    # as the test firmware.
+    if len(ctx.attr.srcs) == 0 and len(ctx.attr.deps) == 0 and ctx.attr.binary:
+        if exec_env.provider in ctx.attr.binary:
+            p = ctx.attr.binary[exec_env.provider]
+        elif DefaultInfo in ctx.attr.binary:
+            files = ctx.attr.binary[DefaultInfo].files.to_list()
+            if len(files) != 1:
+                fail("Expected `binary` to provide exactly one file")
+            p = exec_env.provider(default = files[0])
+        else:
+            fail("No files provided by `binary`")
+    elif (ctx.attr.srcs or ctx.attr.deps) and ctx.attr.binary:
+        fail("Expected either srcs+deps or a binary.")
```

Comment:
This error message is unclear. I suggest `Expected a binary and either srcs or deps`.

Reply:
That's not quite right either.  The good condition is `binary xor (srcs or deps)`.

How about "Expected either a binary or srcs/deps"?

Reply:
I've changed direction on this.  There are some tests that need to use multiple binaries or compile from source _and_ interact with a binary.  I've changed the test rules to accept any number of test binaries.

Reply:
The updated approach LGTM.